File: 464.我能赢吗.cpp
/*
 * @lc app=leetcode.cn id=464 lang=cpp
 *
 * [464] 我能赢吗
 */

// @lc code=start
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    // 完全写错，你写成了先手能获得的最大整数值，但这是显然的
    // int dp_order(int i, int j, vector<vector<int>> a, bool first){
    //     //当次是否先手，上一次是否为偶数（mod2整除）
    //     if(first){
    //         return a[i][j];
    //     }else{
    //         return (i+j)*(j-i+1)/2-a[i][j];
    //     }
    // }

    // bool canIWin(int maxChoosableInteger, int desiredTotal) {
    //     int n=maxChoosableInteger;
    //     // int a[n+1][n+1]={};
    //     vector<vector<int>> dp (n+1,vector<int>(n+1,0)); // 初始化为0，大小是否合适
    //     for(int i=1;i<=n;i++)   dp[i][i]  =i;
    //     for(int i=1;i<n;i++)    dp[i][i+1]=i+1;

    //     for(int interval_len=1;interval_len<=n;interval_len++){ // [1,1]的长度为1
    //         for(int i=1;i<=n-interval_len+1;i++){ // 当前区间是[i, i+len-1]
    //             int j=i+interval_len-1;
    //             if(i==j||i+1==j) continue;
    //             int tmp_score = -1;
    //             for(int mid=i+1;mid<=j-1;mid++){
    //                 // if(mid==i||mid==j){}else{ 不需要处理这种情况。什么时候需要处理
    //                     int a1=dp[i][mid-1]+dp_order(mid,j,dp,(mid-i)%2==0);
    //                     int a2=dp[mid+1][j]+dp_order(i,mid,dp,(j-mid)%2==0);
    //                     int a3=mid;
    //                     a3+=min(dp_order(i,mid-1,dp,(1)%2==0) + dp_order(mid+1,j,dp,(mid-i+1)%2==0),
    //                             dp_order(mid+1,j,dp,(1)%2==0) + dp_order(i,mid-1,dp,(j-mid+1)%2==0));
    //                     int score = max(a1,max(a2,a3)); // 
    //                     if(score>tmp_score){
    //                         tmp_score = score;
    //                     }
    //                 // }
    //             }
    //             dp[i][j]=tmp_score;
    //             // cout<<"["<<i<<", "<<j<<"]: "<<dp[i][j]<<", ";
    //         }
    //     }
    //     for(int i=1;i<=n;i++){
    //         for(int j=i;j<=n;j++){
    //             cout<<"["<<i<<", "<<j<<"]: "<<dp[i][j]<<", ";
    //         }
    //         cout<<endl;
    //     }
    //     return dp[1][n]>=desiredTotal;
    // }
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
        
    }

};
// @lc code=end
// 累计整数和

int main() {
    // TODO: Write your C++ code here
    cout<<Solution().canIWin(5,9);
    return 0;
}

File: 动态规划/464.我能赢吗.py
#
# @lc app=leetcode.cn id=464 lang=python
# @lcpr version=30120
#
# [464] 我能赢吗
#


# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
visited = [-1 for _ in range(1<<20)]
# -1表示没有，1表示true，0表示false

class Solution(object):
    def dfs(self, maxChoosableInteger, desiredTotal, state, nowScore):
        # 递归的状态只和state有关，nowscore可以由state计算，只是因为方便所以在dfs传递
        # dfs先查看visited数组。算出结果是更新visited数组

        if visited[state]!=-1:
            return visited[state]==1 # 不能不要==1，否则返回的就是1(int)而不是true(bool)
        for i in range(1,maxChoosableInteger+1):
            if 1<<(i-1) & state:    # 位操作方法和c++相同
                # 已经被选择，跳过
                continue
            elif i+nowScore>=desiredTotal:
                visited[state]=1
                return True
            else:
                if self.dfs(maxChoosableInteger, desiredTotal, state|1<<(i-1), nowScore+i)==False:
                # 为什么显示dfs未定义。python中需要使用self.，并且在参数中声明self
                    visited[state]=1
                    return True
        visited[state]=0
        return False

    # 首先应该想到，本题是一道搜索题
    def canIWin(self, maxChoosableInteger, desiredTotal):
        """
        :type maxChoosableInteger: int
        :type desiredTotal: int
        :rtype: bool
        """
        if (1+maxChoosableInteger)*maxChoosableInteger/2 <desiredTotal:
            return False
            # 虽然后手的人会输，但是先手的也没法赢
        
        # 力口的判题策略是相同类生命一次，但是方法调用多次。因此使用全局变量需要重新初始化
        # 全局变量在函数中使用需要global，否则相当于重新定义变量
        global visited
        visited = [-1 for _ in range(1<<20)]
        return self.dfs(maxChoosableInteger, desiredTotal, 0, 0)
        # 直接写会有部分测试样例超时，需要进行记忆化搜索 
# @lc code=end



#
# @lcpr case=start
# 10\n11\n
# @lcpr case=end

# @lcpr case=start
# 10\n0\n
# @lcpr case=end

# @lcpr case=start
# 10\n1\n
# @lcpr case=end

#

if __name__ == "__main__":
    # TODO: Write your Python code here
    print(Solution().canIWin(5, 50))
